cleverhans/attacks.py

```diff
@@ -1,4 +1,313 @@
+from abc import ABCMeta, abstractmethod
 import numpy as np
+import warnings
+
+from .utils import random_targets
+
+
+class Attack:
+    """
+    Abstract base class for all attack classes. All attacks must override the
+    `craft` method that returns adversarial examples corresponding to the input
+    data. However, the `generate_symbolic` method is optional.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, x, pred, back='tf', sess=None, clip_min=None,
+                 clip_max=None, params={}):
+        """
+        :param x: The model's symbolic inputs.
```

Comment:
Why is `x` a *required* argument to the constructor?

If it's optional to implement `generate_symbolic`, then this should be optional too. If there's no `generate_symbolic`, this will never be used.